suite/40-mrouter.sh

```diff
@@ -77,104 +104,138 @@ mrouter_report_one()
 mrouter_inject_and_report()
 {
     local all_grps="$1"
-    local b2_grps="$2"
-    local b3_grps="$3"
-    local b3_mrouter=$4
+    local out_grps="$2"
+    local rport_grps="$3"
+    local rport_mrouter=$4
 
-    capture -f "icmp or icmp6 or ether proto 0xbbbb" $h2 $h3
+    capture -f "icmp or icmp6 or ether proto 0xbbbb" $oh $rh
 
-    step "  Inject MAC/IPv4/IPv6 multicast to all groups ($all_grps) on $h1"
+    step "  Inject MAC/IP multicast to all groups ($all_grps) on $ih"
     for g in $all_grps; do
-	eth -g $g | { cat; echo U from $h1; } | inject $h1
-	mcast_gen -c 1 -g $(ip4_grp $g) $ip4_src
-	mcast_gen -c 1 -g $(ip6_grp $g)%$h1 $ip6_src
+	eth -g $g | { cat; echo U from $ih; } | inject $ih
+	mcast_gen -c 1 -g $(mrouter_ip4_grp $g) $mrouter_ip4_src
+
+	# TODO: Nemesis can't yet generate MLD queries, so we only
+	# test IPv6 for the port version of the test, i.e. when the
+	# bridge is generating queries internally.
+	if [ $mode = port ]; then
+	    mcast_gen -c 1 -g $(mrouter_ip6_grp $g)%$ih $mrouter_ip6_src
+	fi
     done
 
-    mrouter_report_one $h2 no "$b2_grps" "$all_grps" || return 1
-    mrouter_report_one $h3 $b3_mrouter "$b3_grps" "$all_grps" || return 1
+    mrouter_report_one $oh no "$out_grps" "$all_grps" || return 1
+    mrouter_report_one $rh $rport_mrouter "$rport_grps" "$all_grps" || return 1
 }
 
-mrouter_port()
+mrouter_test()
 {
     require3loops
 
-    local in=$b1
-    local out=$b2
-    local rport=$b3
+    mode=$1
 
-    local bropts="mcast_snooping 1"
+    ib=$b1
+    ih=$h1
+    ob=$b2
+    oh=$h2
+    qb=$([ $mode = host ] && echo $b3 || echo $br0)
+    qh=$([ $mode = host ] && echo $h3)
+    rb=$([ $mode = port ] && echo $b3 || echo $br0)
+    rh=$([ $mode = port ] && echo $h3 || echo $br0)
+
+    local bropts="mcast_snooping 1 mcast_flood_mrouters_only 0"
```

Comment:
`mcast_flood_mrouters_only` is an internal patch. Maybe add a check if it exists and skip the test if not, or output a warning and create the bridge without it. Now it will just fail to create the bridge if not present. Since it only affects when running in software it would be fine if it allows the test to still run when testing hardware.

Reply:
Right, I keep forgetting that. This was part of a series from @troglobit, of which only half was merged upstream IIRC. I'll work out a better fallback.

Thanks!

Reply:
I was great that you pointed this out. It got me to dig a bit deeper, and it will most likely result in us dropping this patch and just handle flooding in the way that the upstream bridge does.

The test now expects the behavior corresponding to `mcast_flood_mrouters_only=1`, _but_, it accepts (with a warning) systems that behave as if `mcast_flood_mrouters_only=0` (e.g. offloaded bridges on an mv88e6xxx system).

Fair?

Reply:
Well done! Showing warnings seems like the best way to solve this issue since it allows proper testing of all the cases.